### PR TITLE
AVR: implement i2c interface

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds the AVR implementation for the i2c interface, along with an example using a BlinkM i2c RGB LED.

### CODE REVIEW

File: src/machine/machine_avr.go [hunk 1]
```diff
@@ -157,3 +157,96 @@ func (a ADC) Get() uint16 {
...
+	*avr.TWCR = avr.TWCR_TWEN
+}
+
+// I2CStart starts a communication session.
+func I2CStart() {
+	// Clear TWI interrupt flag, put start condition on SDA, and enable TWI.
+	*avr.TWCR = (avr.TWCR_TWINT | avr.TWCR_TWSTA | avr.TWCR_TWEN)
+
+	// Wait till start condition is transmitted.
+	for ok := true; ok; ok = (*avr.TWCR & avr.TWCR_TWINT) == 0 {
```
Comment (aykevl) [inline src/machine/machine_avr.go]: Why not just `for (*avr.TWCR & avr.TWCR_TWINT) == 0 {`?
Also, calling it `ok` is a bit misleading: as long as `ok` is true, the condition is not transmitted (it really means 'not ok').

(This also applies to similar statements below).

File: src/machine/machine_avr.go [hunk 2]
```diff
@@ -157,3 +157,96 @@ func (a ADC) Get() uint16 {
...
+	*avr.PORTC |= (avr.DIDR0_ADC4D | avr.DIDR0_ADC5D)
+
+	// Initialize twi prescaler and bit rate.
+	*avr.TWSR |= (avr.TWSR_TWPS0 | avr.TWSR_TWPS1)
+
+	// twi bit rate formula from atmega128 manual pg. 204:
+	// SCL Frequency = CPU Clock Frequency / (16 + (2 * TWBR))
+	// NOTE: TWBR should be 10 or higher for master mode.
+	// It is 72 for a 16mhz board with 100kHz TWI
+	*avr.TWBR = avr.RegValue(72)
```
Comment (aykevl) [inline src/machine/machine_avr.go]: You can calculate this value at compile time with constants, to avoid hardcoding this value in the I2C layer.
For example, you can add a `CPU_FREQUENCY` (or something like that) to machine_arduino.go with value `16000000`, and use it here to calculate the TWBR frequency.

File: src/machine/machine_avr.go [hunk 3]
```diff
@@ -157,3 +157,96 @@ func (a ADC) Get() uint16 {
...
+	for ok := true; ok; ok = (*avr.TWCR & avr.TWCR_TWINT) == 0 {
+	}
+}
+
+// I2CReadByte reads a single byte from the I2C bus.
+func I2CReadByte() byte {
+	// Clear TWI interrupt flag and enable TWI.
+	*avr.TWCR = (avr.TWCR_TWEN | avr.TWCR_TWINT | avr.TWCR_TWEA)
+
+	// Wait till data is transmitted.
```
Comment (aykevl) [inline src/machine/machine_avr.go]: Copy/paste error? You're waiting until the data is received.

**Comment by deadprogram** [comment]
Updated based on all suggested items. Thanks for the feedback.

**Comment by deadprogram** [comment]
The wiki has an error for the new I2C interface definition. The method `ReadByte(data byte)` should be `ReadByte() byte`.

File: src/examples/blinkm/blinkm.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	machine.I2C0.ReadFrom(0x09, version)
+	println("Firmware version:", string(version[0]), string(version[1]))
+
+	count := 0
+	for {
+		machine.I2C0.WriteTo(0x09, []byte("n"))
+
+		switch count {
+		case 0:
+			machine.I2C0.WriteTo(0x09, []byte{0x99, 0xff, 0x00})
```
Comment (aykevl) [inline src/examples/blinkm/blinkm.go]: Suggestion: add a comment what color you're sending with these commands (I'm guessing yellow in this case?).

File: src/machine/machine_avr.go [hunk 5]
```diff
@@ -157,3 +159,119 @@ func (a ADC) Get() uint16 {
 	high := uint16(*avr.ADCH)
 	return uint16(low) | uint16(high<<8)
 }
+
+// I2C on the Arduino
+type I2C struct {
+	// Cannot have just an empty struct or compile optimization fails.
+	Bus uint8
```
Comment (aykevl) [inline src/machine/machine_avr.go]: Ouch, this is one reason why I was adding tests yesterday - I was afraid something would go wrong with empty structs. Ideally this should be fixed before merging. It's a good workaround but should not have been necessary.

File: src/machine/machine_avr.go [hunk 6]
```diff
@@ -157,3 +159,119 @@ func (a ADC) Get() uint16 {
...
+	i2c.WriteByte(address<<1 + 1)
+
+	for i, _ := range data {
+		data[i] = i2c.ReadByte()
+	}
+
+	i2c.Stop()
+}
+
+// CWriteByte writes a single byte to the I2C bus.
```
Comment (aykevl) [inline src/machine/machine_avr.go]: typo?

File: src/machine/machine_avr.go [hunk 7]
```diff
@@ -6,6 +6,8 @@ import (
 	"device/avr"
 )
 
+const CPU_FREQUENCY = 16000000
```
Comment (aykevl) [inline src/machine/machine_avr.go]: Can you move this to machine_arduino.go? This value is board-specific, and should be defined separately for each board (for example, some Arduino variants like the [SparkFun Pro Micro](https://www.sparkfun.com/products/12587) run at 8MHz).

File: src/examples/blinkm/blinkm.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
+// Connects to an BlinkM I2C RGB LED.
```
Comment (aykevl) [inline src/examples/blinkm/blinkm.go]: Suggestion: add a link to the specification for the BlinkM I2C protocol.

**Comment by aykevl** [comment]
> The wiki has an error for the new I2C interface definition

You are completely right. Fixed it.

File: src/examples/blinkm/blinkm.go (hunk 4 above)
Comment (deadprogram) [reply src/examples/blinkm/blinkm.go]: Actually I had just made up some numbers. I will chose a real color.

File: src/machine/machine_avr.go (hunk 6 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: Correct.

File: src/examples/blinkm/blinkm.go (hunk 8 above)
Comment (deadprogram) [reply src/examples/blinkm/blinkm.go]: Done.

File: src/machine/machine_avr.go (hunk 7 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: I think you mean `board_arduino.go`, and done.

File: src/machine/machine_avr.go (hunk 6 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: Fixed.

**Comment by deadprogram** [comment]
OK, I think we're good now.

**Comment by aykevl** [comment]
The fix turned out to be relatively easy and is now in master.

**Comment by aykevl** [comment]
I've tested it and it compiles without the two lines related to `I2C.Bus`. The output is exactly the same as with the two lines (as expected). Are you okay with me removing those two lines and squashing+merging the result?

**Comment by deadprogram** [comment]
Do it! :balloon:

**Comment by aykevl** [comment]
Merged in 12fb4f3f91a10d7b6c8af0bda4a41cac34af5395. Thank you for the work!
